File: Core/src/main/java/com/cyrillrx/utils/Serializer.java
package com.cyrillrx.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

/**
 * Serialization helper that uses {@link Gson} implementations to serialize and deserialize objects.
 *
 * @author Cyril Leroux
 *         Created 31/03/2017.
 */
public class Serializer {

    private static final Gson DEFAULT;
    private static final Gson PRETTY_PRINT;
    private static final Gson NO_HTML_ESCAPING;

    static {
        DEFAULT = new Gson();

        PRETTY_PRINT = new GsonBuilder()
                .disableHtmlEscaping()
                .setPrettyPrinting()
                .serializeNulls()
                .create();

        NO_HTML_ESCAPING = new GsonBuilder()
                .disableHtmlEscaping()
                .create();
    }

    /** Serializes the given object with the default {@link Gson} implementation. */
    public String serialize(Object object) { return DEFAULT.toJson(object); }

    public static <T> T deserialize(String json, Class<T> clazz) {
        return DEFAULT.fromJson(json, clazz);
    }

    /** Serializes the given object without escaping characters such as < > etc. */
    public String serializeNoEscaping(Object object) { return NO_HTML_ESCAPING.toJson(object); }

    /** Serializes and pretty prints the given object. */
    public String prettyPrint(Object object) { return PRETTY_PRINT.toJson(object); }

}

File: core/src/main/java/com/cyrillrx/utils/Serializer.kt
package com.cyrillrx.utils

import com.google.gson.Gson
import com.google.gson.GsonBuilder

/**
 * Serialization helper that uses [Gson] implementations to serialize and deserialize objects.
 *
 * @author Cyril Leroux
 *         Created on 31/03/2017
 */

val defaultSerializer: Gson = Gson()

val prettyPrintSerializer: Gson = GsonBuilder()
        .disableHtmlEscaping()
        .setPrettyPrinting()
        .serializeNulls()
        .create()

val noHtmlEscapeSerializer: Gson = GsonBuilder()
        .disableHtmlEscaping()
        .create()

/** Serializes the receiver object with the default [Gson] implementation.  */
fun Any.serialize(): String = defaultSerializer.toJson(this)

fun <T> String.deserialize(clazz: Class<T>): T = defaultSerializer.fromJson(this, clazz)

inline fun <reified T> String.deserialize(): T = deserialize(T::class.java)

/** Serializes the receiver object without escaping characters such as < > etc.  */
fun Any.serializeNoEscaping(): String = noHtmlEscapeSerializer.toJson(this)

inline fun <reified T> String.deserializeNoEscaping(): T =
        noHtmlEscapeSerializer.fromJson(this, T::class.java)

/** Serializes and pretty prints the given object.  */
fun Any.prettyPrint(): String = prettyPrintSerializer.toJson(this)